File: leetcode/cpp/7_reverse_integer.cpp
#include <cmath>
#include <iostream>

class Solution {
 public:
  int reverse(int x) {
    bool flag = x > 0 ? true : false;
    if (!flag) x = -x;
    int n = 0, i, res = 0;
    int m = x;
    while (m) {
      n++;
      m = m / 10;
    }
    m = x;
    for (i = 0; i < n; i++) {
      res += (m % 10) * pow(10, n - i - 1);
      if (res < 0) {  // overflow
        return 0;
      }
      m = m / 10;
    }
    if (flag)
      return res;
    else
      return -res;
  }
};

int main(int argc, char *argv[]) {
  Solution s;
  std::cout << s.reverse(-123) << "\n";
  std::cout << s.reverse(3) << "\n";
  std::cout << s.reverse(0) << "\n";
  return 0;
}

File: leetcode/go/src/leetcode/7_reverse_integer.go
package leetcode

func reverse(x int) int {
	if x < 0 {
		return -reverse(-x)
	}
	var res int64
	for x > 0 {
		res = res*10 + int64(x)%10
		x = x / 10
	}
	if res > (1<<31 - 1) {
		return 0
	}
	return int(res)
}
